File: src/scanner.c
#include <stdbool.h>
#include <string.h>

#include "scanner.h"

void scanner_init(Scanner *scanner, const char *source) {
  scanner->start = source;
  scanner->current = source;
  scanner->line = 1;
}

static bool at_eof(Scanner *scanner) { return *scanner->current == '\0'; }

static Token make_token(Scanner *scanner, TokenType type) {
  Token token;
  token.type = type;
  token.start = scanner->start;
  token.length = (int)(scanner->current - scanner->start);
  token.line = scanner->line;
  return token;
}

static Token make_error(Scanner *scanner, const char *msg) {
  Token token;
  token.type = TOKEN_ERROR;
  token.start = msg;
  token.length = (int)(strlen(msg));
  token.line = scanner->line;
  return token;
}

static char advance(Scanner *scanner) {
  scanner->current++;
  return scanner->current[-1];
}

static char peek(Scanner *scanner) { return *scanner->current; }

static char peek_next(Scanner *scanner) {
  if (at_eof(scanner)) {
    return '\0';
  }
  return scanner->current[1];
}

static bool match(Scanner *scanner, const char expected) {
  if (at_eof(scanner)) {
    return false;
  }

  if (*scanner->current != expected) {
    return false;
  }

  scanner->current++;
  return true;
}

static void skip_ignored(Scanner *scanner) {
  for (;;) {
    char c = peek(scanner);
    switch (c) {

    // Whitespace
    case ' ':
    case '\r':
    case '\t':
      advance(scanner);
      break;
    case '\n':
      scanner->line++;
      advance(scanner);
      break;

      // Comments
    case '/':
      if (peek_next(scanner) == '/') {
        while (peek(scanner) != '\n' && !at_eof(scanner)) {
          advance(scanner);
        }
      } else {
        return;
      }
      break;

    default:
      return;
    }
  }
}

static Token scan_string(Scanner *scanner) {
  while (peek(scanner) != '"' && !at_eof(scanner)) {
    if (peek(scanner) == '\n') {
      scanner->line++;
    }
    advance(scanner);
  }

  if (at_eof(scanner)) {
    return make_error(scanner, "unterminated string");
  }

  // Consume the '"' found by peek()
  advance(scanner);
  return make_token(scanner, TOKEN_STRING);
}

static bool is_digit(const char c) { return '0' <= c && c <= '9'; }

static Token scan_number(Scanner *scanner) {
  while (is_digit(peek(scanner))) {
    advance(scanner);
  }

  // Only consume the dot if there are also more digits afterward. Otherwise, it
  // indicates a method call rather than a decimal number.
  if (peek(scanner) == '.' && is_digit(peek_next(scanner))) {
    advance(scanner); // '.'

    // Everything after the dot
    while (is_digit(peek(scanner))) {
      advance(scanner);
    }
  }

  return make_token(scanner, TOKEN_NUMBER);
}

static bool is_alpha(const char c) {
  return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || c == '_';
}

static TokenType keyword_or_identifier(Scanner *scanner) {
  // The chars tracked by the scanner _are not_ a null-terminated string. Be
  // careful with lengths and comparisons!
  const char *id = scanner->start;
  size_t len = (size_t)(scanner->current - scanner->start);

#define KWD_IF(kwd, type)                                                      \
  if (strlen(kwd) == len && strncmp(kwd, id, len) == 0) {                      \
    return type;                                                               \
  }

  KWD_IF("and", TOKEN_AND)
  KWD_IF("class", TOKEN_CLASS)
  KWD_IF("else", TOKEN_ELSE)
  KWD_IF("false", TOKEN_FALSE)
  KWD_IF("if", TOKEN_IF)
  KWD_IF("nil", TOKEN_NIL)
  KWD_IF("or", TOKEN_OR)
  KWD_IF("print", TOKEN_PRINT)
  KWD_IF("return", TOKEN_RETURN)
  KWD_IF("super", TOKEN_SUPER)
  KWD_IF("true", TOKEN_TRUE)
  KWD_IF("var", TOKEN_VAR)
  KWD_IF("while", TOKEN_WHILE)

#undef KWD

  return TOKEN_IDENTIFIER;
}

static Token scan_identifier(Scanner *scanner) {
  while (is_alpha(peek(scanner)) || is_digit(peek(scanner))) {
    advance(scanner);
  }

  TokenType type = keyword_or_identifier(scanner);
  return make_token(scanner, type);
}

Token scanner_next(Scanner *scanner) {
  skip_ignored(scanner);

  scanner->start = scanner->current;

  if (at_eof(scanner)) {
    return make_token(scanner, TOKEN_EOF);
  }

  char c = advance(scanner);

  if (is_alpha(c)) {
    return scan_identifier(scanner);
  }
  if (is_digit(c)) {
    return scan_number(scanner);
  }

  switch (c) {
  // One character, no lookahead
  case '(':
    return make_token(scanner, TOKEN_LEFT_PAREN);
  case ')':
    return make_token(scanner, TOKEN_RIGHT_PAREN);
  case '{':
    return make_token(scanner, TOKEN_LEFT_BRACE);
  case '}':
    return make_token(scanner, TOKEN_RIGHT_BRACE);
  case ';':
    return make_token(scanner, TOKEN_SEMICOLON);
  case ',':
    return make_token(scanner, TOKEN_COMMA);
  case '.':
    return make_token(scanner, TOKEN_DOT);
  case '-':
    return make_token(scanner, TOKEN_MINUS);
  case '+':
    return make_token(scanner, TOKEN_PLUS);
  case '/':
    return make_token(scanner, TOKEN_SLASH);
  case '*':
    return make_token(scanner, TOKEN_STAR);

  // One or two characters, 1 lookahead
  case '!': {
    TokenType type = match(scanner, '=') ? TOKEN_BANG_EQUAL : TOKEN_BANG;
    return make_token(scanner, type);
  }
  case '=': {
    TokenType type = match(scanner, '=') ? TOKEN_EQUAL_EQUAL : TOKEN_EQUAL;
    return make_token(scanner, type);
  }
  case '<': {
    TokenType type = match(scanner, '=') ? TOKEN_LESS_EQUAL : TOKEN_LESS;
    return make_token(scanner, type);
  }
  case '>': {
    TokenType type = match(scanner, '=') ? TOKEN_GREATER_EQUAL : TOKEN_GREATER;
    return make_token(scanner, type);
  }

  case '"':
    return scan_string(scanner);

  default:
    return make_error(scanner, "unexpected character");
  }
}

File: src/scanner.rs
#[derive(Debug, Copy, Clone, Default)]
#[repr(C)]
pub struct Token<'a> {
    pub(crate) r#type: TokenType,
    pub(crate) text: &'a str,
    pub(crate) line: usize,
}

impl Token<'_> {
    pub(crate) fn synthetic(text: &'static str) -> Self {
        let ty = match text {
            "super" => TokenType::Super,
            "this" => TokenType::This,
            _ => unreachable!(),
        };

        Self {
            r#type: ty,
            line: 0,
            text,
        }
    }

    pub(crate) fn text(&self) -> &str {
        self.text
    }
}

#[repr(C)]
#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, Hash)]
pub enum TokenType {
    // Single-character tokens
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // One- or two-character tokens
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals
    Identifier,
    String,
    Number,

    // Keywords
    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    // Synthetic tokens
    #[default]
    Error,
    Eof,
}

fn is_alpha(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_digit(c: char) -> bool {
    c.is_ascii_digit()
}

fn keyword_or_identifier(text: &str) -> TokenType {
    macro_rules! kwd_if {
        ($kwd:expr, $ty:expr) => {
            if text == $kwd {
                return $ty;
            }
        };
    }

    kwd_if!("and", TokenType::And);
    kwd_if!("class", TokenType::Class);
    kwd_if!("else", TokenType::Else);
    kwd_if!("false", TokenType::False);
    kwd_if!("for", TokenType::For);
    kwd_if!("fun", TokenType::Fun);
    kwd_if!("if", TokenType::If);
    kwd_if!("nil", TokenType::Nil);
    kwd_if!("or", TokenType::Or);
    kwd_if!("print", TokenType::Print);
    kwd_if!("return", TokenType::Return);
    kwd_if!("super", TokenType::Super);
    kwd_if!("this", TokenType::This);
    kwd_if!("true", TokenType::True);
    kwd_if!("var", TokenType::Var);
    kwd_if!("while", TokenType::While);

    TokenType::Identifier
}

#[repr(C)]
pub struct Scanner<'source> {
    source: &'source str,
    start: usize,
    current: usize,
    line: usize,
}

impl<'source> Scanner<'source> {
    pub fn new(source: &'source str) -> Self {
        Self {
            source,
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn next_token(&mut self) -> Token<'source> {
        self.skip_ignored();

        self.start = self.current;

        if self.at_eof() {
            return self.make_token(TokenType::Eof);
        }

        let c = self.advance();

        if is_alpha(c) {
            return self.scan_identifier();
        }
        if is_digit(c) {
            return self.scan_number();
        }

        match c {
            // One character, no lookahead
            '(' => self.make_token(TokenType::LeftParen),
            ')' => self.make_token(TokenType::RightParen),
            '{' => self.make_token(TokenType::LeftBrace),
            '}' => self.make_token(TokenType::RightBrace),
            ';' => self.make_token(TokenType::Semicolon),
            ',' => self.make_token(TokenType::Comma),
            '.' => self.make_token(TokenType::Dot),
            '-' => self.make_token(TokenType::Minus),
            '+' => self.make_token(TokenType::Plus),
            '/' => self.make_token(TokenType::Slash),
            '*' => self.make_token(TokenType::Star),

            // One or two characters, 1 lookahead
            '!' => {
                if self.match_('=') {
                    self.make_token(TokenType::BangEqual)
                } else {
                    self.make_token(TokenType::Bang)
                }
            }
            '=' => {
                if self.match_('=') {
                    self.make_token(TokenType::EqualEqual)
                } else {
                    self.make_token(TokenType::Equal)
                }
            }
            '<' => {
                if self.match_('=') {
                    self.make_token(TokenType::LessEqual)
                } else {
                    self.make_token(TokenType::Less)
                }
            }
            '>' => {
                if self.match_('=') {
                    self.make_token(TokenType::GreaterEqual)
                } else {
                    self.make_token(TokenType::Greater)
                }
            }

            '"' => self.scan_string(),

            _ => self.make_error("Unexpected character."),
        }
    }

    fn skip_ignored(&mut self) {
        loop {
            match self.peek() {
                // Whitespace
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }

                // Comments
                '/' => {
                    if self.peek_next() == '/' {
                        while self.peek() != '\n' && !self.at_eof() {
                            self.advance();
                        }
                    } else {
                        return;
                    }
                }

                _ => return,
            }
        }
    }

    fn scan_identifier(&mut self) -> Token<'source> {
        while is_alpha(self.peek()) || is_digit(self.peek()) {
            self.advance();
        }

        let text = &self.source[self.start..self.current];

        let ty = keyword_or_identifier(text);
        self.make_token(ty)
    }

    fn scan_number(&mut self) -> Token<'source> {
        while is_digit(self.peek()) {
            self.advance();
        }

        // Only consume the dot if there are also more digits afterward. Otherwise, it indicates a
        // method call rather than a decimal number.
        if self.peek() == '.' && is_digit(self.peek_next()) {
            self.advance(); // `.`

            while is_digit(self.peek()) {
                self.advance();
            }
        }

        self.make_token(TokenType::Number)
    }

    fn scan_string(&mut self) -> Token<'source> {
        while self.peek() != '"' && !self.at_eof() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.at_eof() {
            return self.make_error("Unterminated string.");
        }

        self.advance(); // peek `"`
        self.make_token(TokenType::String)
    }

    fn at_eof(&self) -> bool {
        self.peek() == '\0'
    }

    fn peek(&self) -> char {
        self.source[self.current..]
            .chars()
            .next()
            .unwrap_or_default()
    }

    fn peek_next(&self) -> char {
        if self.at_eof() {
            return '\0';
        }

        let mut chars = self.source[self.current..].chars();
        chars.next().unwrap(); // peek 0
        chars.next().unwrap_or_default() // peek 1
    }

    fn advance(&mut self) -> char {
        let c = self.peek();
        self.current += c.len_utf8();
        c
    }

    fn match_(&mut self, expected: char) -> bool {
        if self.at_eof() {
            return false;
        }

        if self.peek() != expected {
            return false;
        }

        self.advance();
        true
    }

    fn make_token(&self, ty: TokenType) -> Token<'source> {
        let text = &self.source[self.start..self.current];

        Token {
            r#type: ty,
            text,
            line: self.line,
        }
    }

    fn make_error<'msg>(&self, msg: &'msg str) -> Token<'msg> {
        Token {
            r#type: TokenType::Error,
            text: msg,
            line: self.line,
        }
    }
}
